miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,458 @@
+    /// ```
+    /// And keeps only trees *after* that bit, meaning that the tree at `tree_idx` is also removed,
+    /// resulting in `0b0101_0100`.
+    ///
+    /// ```
+    /// # use miden_crypto::merkle::Forest;
+    /// let range = Forest::new(0b0101_0110);
+    /// assert_eq!(range.trees_larger_than(1), Forest::new(0b0101_0100));
+    /// ```
+    pub fn trees_larger_than(self, tree_idx: u32) -> Self {
+        self & high_bitmask(tree_idx + 1)
+    }
+
+    /// Creates a new forest with all possible trees smaller than the smallest tree in this
+    /// forest.
+    ///
+    /// This forest must have exactly one tree.
+    ///
+    /// # Panics
+    /// With debug assertions enabled, this function panics if this forest does not have
+    /// exactly one tree.
+    ///
+    /// For a non-panicking version of this function, see [`Forest::all_smaller_trees()`].
+    pub fn all_smaller_trees_unchecked(self) -> Self {
+        debug_assert_eq!(self.num_trees(), 1);
+        Self::new(self.0 - 1)
+    }
+
+    /// Creates a new forest with all possible trees smaller than the smallest tree in this
+    /// forest, or returns `None` if this forest has more or less than one tree.
+    ///
+    /// If the forest cannot have more or less than one tree, use
+    /// [`Forest::all_smaller_trees_unchecked()`] for performance.
+    pub fn all_smaller_trees(self) -> Option<Forest> {
+        if self.num_trees() != 1 {
+            return None;
+        }
+        Some(self.all_smaller_trees_unchecked())
+    }
+
+    /// Returns a forest with exactly one tree, one size (depth) larger than the current one.
+    pub fn next_larger_tree(self) -> Self {
+        debug_assert_eq!(self.num_trees(), 1);
+        Forest(self.0 << 1)
+    }
+
+    /// Returns true if the forest contains a single-node tree.
+    pub fn has_single_leaf_tree(self) -> bool {
+        self.0 & 1 != 0
+    }
+
+    /// Add a single-node tree if not already present in the forest.
+    pub fn single_leaf_tree_added(self) -> Self {
+        Self::new(self.0 | 1)
+    }
+
+    /// Remove the single-node tree if present in the forest.
+    pub fn single_leaf_tree_removed(self) -> Self {
+        Self::new(self.0 & (usize::MAX - 1))
+    }
+
+    /// Returns a new forest that does not have the trees that `other` has.
+    pub fn without_trees(self, other: Forest) -> Self {
+        self ^ other
+    }
+
+    /// Returns index of the forest tree for a specified leaf index.
+    pub fn tree_index(&self, leaf_idx: usize) -> usize {
+        let root = self
+            .leaf_to_corresponding_tree(leaf_idx)
+            .expect("position must be part of the forest");
+        let smaller_tree_mask = Self::new(2_usize.pow(root) - 1);
+        let num_smaller_trees = (*self & smaller_tree_mask).num_trees();
+        self.num_trees() - num_smaller_trees - 1
+    }
+
+    /// Returns the in-order index of the root element.
+    pub fn root_in_order_index(&self) -> InOrderIndex {
```

Comment:
What does "root element" mean here? It may be worth expanding the doc comments here a bit.

Reply:
It's mostly a copy of https://github.com/0xMiden/crypto/blob/7e534bdf4072f3332fa18528d2567e4d059df465/miden-crypto/src/merkle/mmr/partial.rs#L615-L630, but I can clarify the docs a bit.